File: backend_python/app/clients/llm/ollama_adapter.py
"""Ollama LLM adapter."""
import os
import json
from typing import Optional
import httpx
from .openai_adapter import LLMMessage, LLMOptions, LLMResponse


async def generate_with_ollama(options: LLMOptions) -> LLMResponse:
    """Generate completion using Ollama."""
    base_url = os.getenv("LLM_BASE_URL", "http://localhost:11434")
    model = options.model or os.getenv("LLM_MODEL", "phi3")

    async with httpx.AsyncClient(timeout=120.0) as client:
        response = await client.post(
            f"{base_url}/api/chat",
            json={
                "model": model,
                "messages": [msg.to_dict() for msg in options.messages],
                "stream": False,
                "options": {
                    "temperature": options.temperature or 0.7,
                    "num_predict": options.max_tokens or 500,
                    "num_ctx": 2048,
                    "top_k": 20,
                    "top_p": 0.9,
                    "repeat_penalty": 1.1,
                    "num_thread": 2,
                    "numa": False,
                },
            },
        )

        if response.status_code != 200:
            raise ValueError(f"Ollama API error: {response.status_code} - {response.text}")

        # Ollama may return streaming JSON (multiple lines) even with stream:false
        text = response.text
        try:
            # Try parsing as single JSON first
            data = response.json()
        except json.JSONDecodeError:
            # If it's streaming format (multiple JSON lines), parse all chunks
            lines = [line.strip() for line in text.strip().split("\n") if line.strip()]
            if not lines:
                raise ValueError("Ollama returned empty response")

            # Accumulate content from all chunks
            full_content = ""
            final_data = None
            model_name = model
            prompt_tokens = 0
            completion_tokens = 0

            for line in lines:
                try:
                    chunk = json.loads(line)
                    if chunk.get("message", {}).get("content"):
                        full_content += chunk["message"]["content"]
                    if chunk.get("model"):
                        model_name = chunk["model"]
                    if chunk.get("prompt_eval_count"):
                        prompt_tokens = chunk["prompt_eval_count"]
                    if chunk.get("eval_count"):
                        completion_tokens = chunk["eval_count"]
                    if chunk.get("done"):
                        final_data = chunk
                except json.JSONDecodeError:
                    continue

            # Use final chunk if available
            if final_data:
                data = {
                    **final_data,
                    "message": {"content": full_content or final_data.get("message", {}).get("content", "")},
                    "model": model_name,
                    "prompt_eval_count": prompt_tokens,
                    "eval_count": completion_tokens,
                }
            elif full_content:
                data = {
                    "message": {"content": full_content},
                    "model": model_name,
                    "prompt_eval_count": prompt_tokens,
                    "eval_count": completion_tokens,
                }
            else:
                # Last resort: try to parse the last line
                data = json.loads(lines[-1])

    return LLMResponse(
        content=data.get("message", {}).get("content", ""),
        model=data.get("model", model),
        usage={
            "prompt_tokens": data.get("prompt_eval_count"),
            "completion_tokens": data.get("eval_count"),
            "total_tokens": (data.get("prompt_eval_count") or 0) + (data.get("eval_count") or 0),
        },
    )


async def check_ollama_health() -> bool:
    """Check if Ollama is available."""
    try:
        base_url = os.getenv("LLM_BASE_URL", "http://localhost:11434")
        async with httpx.AsyncClient(timeout=3.0) as client:
            response = await client.get(f"{base_url}/api/tags")
            return response.status_code == 200
    except Exception:
        return False

File: backend/src/clients/llm/ollamaAdapter.ts
import { LLMMessage, LLMOptions, LLMResponse } from './openaiAdapter';

export async function generateWithOllama(
  options: LLMOptions
): Promise<LLMResponse> {
  const baseUrl = process.env.LLM_BASE_URL || 'http://localhost:11434';
  const model = options.model || process.env.LLM_MODEL || 'phi3';

  const response = await fetch(`${baseUrl}/api/chat`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      model,
      messages: options.messages.map((msg) => ({
        role: msg.role,
        content: msg.content,
      })),
      stream: false, // Request non-streaming response
      options: {
        temperature: options.temperature ?? 0.7, // Balanced for lightweight model
        num_predict: options.max_tokens || 500, // Reduced for lighter model
        num_ctx: 2048, // Reduced context window for lighter model (was 8192)
        top_k: 20, // Reduced for lighter model
        top_p: 0.9, // Slightly reduced for lighter model
        repeat_penalty: 1.1, // Prevent repetition
        num_thread: 2, // Reduced threads for lighter system load (was 8)
        numa: false, // Disable NUMA for lighter systems
      },
    }),
  });

  if (!response.ok) {
    const errorText = await response.text();
    throw new Error(`Ollama API error: ${response.status} - ${errorText}`);
  }

  // Ollama may return streaming JSON (multiple lines) even with stream:false
  // Handle both single JSON and streaming formats
  const text = await response.text();
  let data: {
    message?: { content?: string };
    model?: string;
    prompt_eval_count?: number;
    eval_count?: number;
    done?: boolean;
  };

  try {
    // Try parsing as single JSON first
    data = JSON.parse(text);
  } catch (parseError) {
    // If it's streaming format (multiple JSON lines), parse all chunks
    const lines = text.trim().split('\n').filter(line => line.trim());
    if (lines.length === 0) {
      throw new Error('Ollama returned empty response');
    }

    // Accumulate content from all chunks
    let fullContent = '';
    let finalData: typeof data | null = null;
    let modelName = model;
    let promptTokens = 0;
    let completionTokens = 0;

    for (const line of lines) {
      try {
        const chunk = JSON.parse(line);
        
        // Accumulate message content
        if (chunk.message?.content) {
          fullContent += chunk.message.content;
        }
        
        // Track model and tokens
        if (chunk.model) modelName = chunk.model;
        if (chunk.prompt_eval_count) promptTokens = chunk.prompt_eval_count;
        if (chunk.eval_count) completionTokens = chunk.eval_count;
        
        // Last chunk with done:true is the final one
        if (chunk.done) {
          finalData = chunk;
        }
      } catch (e) {
        // Skip invalid JSON lines
        console.warn('Skipping invalid JSON line from Ollama:', line.substring(0, 50));
      }
    }

    // Use final chunk if available, otherwise construct from accumulated data
    if (finalData) {
      data = {
        ...finalData,
        message: { content: fullContent || finalData.message?.content || '' },
        model: modelName,
        prompt_eval_count: promptTokens,
        eval_count: completionTokens,
      };
    } else if (fullContent) {
      // Fallback: construct response from accumulated content
      data = {
        message: { content: fullContent },
        model: modelName,
        prompt_eval_count: promptTokens,
        eval_count: completionTokens,
      };
    } else {
      // Last resort: try to parse the last line
      try {
        data = JSON.parse(lines[lines.length - 1]);
      } catch (e) {
        throw new Error(`Ollama response parse error: ${text.substring(0, 200)}`);
      }
    }
  }

  return {
    content: data.message?.content || '',
    model: data.model || model,
    usage: {
      prompt_tokens: data.prompt_eval_count,
      completion_tokens: data.eval_count,
      total_tokens: (data.prompt_eval_count || 0) + (data.eval_count || 0),
    },
  };
}

export async function checkOllamaHealth(): Promise<boolean> {
  try {
    const baseUrl = process.env.LLM_BASE_URL || 'http://localhost:11434';
    const response = await fetch(`${baseUrl}/api/tags`, {
      method: 'GET',
      signal: AbortSignal.timeout(3000),
    });
    return response.ok;
  } catch {
    return false;
  }
}
